system/lib/libc/musl/src/locale/newlocale.c

```diff
@@ -1,49 +1,73 @@
 #include <stdlib.h>
 #include <string.h>
+#include <pthread.h>
 #include "locale_impl.h"
-#include "libc.h"
+#include "lock.h"
+
+#define malloc __libc_malloc
+#define calloc undef
+#define realloc undef
+#define free undef
+
+static int default_locale_init_done;
+static struct __locale_struct default_locale, default_ctype_locale;
 
 int __loc_is_allocated(locale_t loc)
 {
-	return loc && loc != C_LOCALE && loc != UTF8_LOCALE;
+	return loc && loc != C_LOCALE && loc != UTF8_LOCALE
+		&& loc != &default_locale && loc != &default_ctype_locale;
 }
 
-locale_t __newlocale(int mask, const char *name, locale_t loc)
+static locale_t do_newlocale(int mask, const char *name, locale_t loc)
 {
-	int i, j;
 	struct __locale_struct tmp;
-	const struct __locale_map *lm;
+
+	for (int i=0; i<LC_ALL; i++) {
+		tmp.cat[i] = (!(mask & (1<<i)) && loc) ? loc->cat[i] :
+			__get_locale(i, (mask & (1<<i)) ? name : "");
+		if (tmp.cat[i] == LOC_MAP_FAILED)
+			return 0;
+	}
 
 	/* For locales with allocated storage, modify in-place. */
 	if (__loc_is_allocated(loc)) {
-		for (i=0; i<LC_ALL; i++)
-			if (mask & (1<<i))
-				loc->cat[i] = __get_locale(i, name);
+		*loc = tmp;
 		return loc;
 	}
 
-	/* Otherwise, build a temporary locale object, which will only
-	 * be instantiated in allocated storage if it does not match
-	 * one of the built-in static locales. This makes the common
-	 * usage case for newlocale, getting a C locale with predictable
-	 * behavior, very fast, and more importantly, fail-safe. */
-	for (j=i=0; i<LC_ALL; i++) {
-		if (loc && !(mask & (1<<i)))
-			lm = loc->cat[i];
-		else
-			lm = __get_locale(i, mask & (1<<i) ? name : "");
-		if (lm) j++;
-		tmp.cat[i] = lm;
+	/* Otherwise, first see if we can use one of the builtin locales.
+	 * This makes the common usage case for newlocale, getting a C locale
+	 * with predictable behavior, very fast, and more importantly, fail-safe. */
+	if (!memcmp(&tmp, C_LOCALE, sizeof tmp)) return C_LOCALE;
+	if (!memcmp(&tmp, UTF8_LOCALE, sizeof tmp)) return UTF8_LOCALE;
+
+	/* And provide builtins for the initial default locale, and a
+	 * variant of the C locale honoring the default locale's encoding. */
+	if (!default_locale_init_done) {
+		for (int i=0; i<LC_ALL; i++)
+			default_locale.cat[i] = __get_locale(i, "");
+		default_ctype_locale.cat[LC_CTYPE] = default_locale.cat[LC_CTYPE];
+		default_locale_init_done = 1;
 	}
+	if (!memcmp(&tmp, &default_locale, sizeof tmp)) return &default_locale;
+	if (!memcmp(&tmp, &default_ctype_locale, sizeof tmp))
+		return &default_ctype_locale;
 
-	if (!j)
-		return C_LOCALE;
-	if (j==1 && tmp.cat[LC_CTYPE]==&__c_dot_utf8)
-		return UTF8_LOCALE;
+	/* If no builtin locale matched, attempt to allocate and copy. */
+	/* XXX EMSCRIPTEN .. */
+	/*if ((loc = malloc(sizeof *loc))) *loc = tmp;*/
 
 	if ((loc = malloc(sizeof *loc))) *loc = tmp;
```

Comment:
Two lines above this (I can't open a comment there as there is a resolved one already) there is the same line but in a comment. Do we want the commented line, or the uncommented one?